src/main/java/com/linkedin/kmf/services/KeyedProduceService.java

```diff
@@ -0,0 +1,123 @@
+
+import static org.apache.kafka.common.utils.Utils.murmur2;
+
+
+public class KeyedProduceService extends ProduceService implements Service {
+  private static final Logger LOG = LoggerFactory.getLogger(KeyedProduceService.class);
+  private static final String METRIC_GROUP_NAME = "keyed-produce-service";
+  private final ConcurrentMap<String, AtomicLong> _nextIndexPerKey;
+
+  public KeyedProduceService(Map<String, Object> props, String name) throws Exception {
+    super(props, name);
+    _nextIndexPerKey = new ConcurrentHashMap<>();
+
+  }
+
+  @Override
+  public void start() {
+    if (_running.compareAndSet(false, true)) {
+      initializeStateForPartitions();
+      _handleNewPartitionsExecutor.scheduleWithFixedDelay(new NewPartitionHandler(), 40_000, 40_000, TimeUnit.MILLISECONDS);
+      LOG.info(_name + "/ProduceService started");
+    }
+  }
+
+  private void initializeStateForPartitions() {
+    Map<Integer, String> keyMapping = generateKeyMappings();
+    int partitionNum = _partitionNum.get();
+    for (int partition = 0; partition < partitionNum; partition++) {
+      String key = keyMapping.get(partition);
+      //This is what preserves sequence numbers across restarts
+      if (!_nextIndexPerKey.containsKey(key)) {
+        _nextIndexPerKey.put(key, new AtomicLong(0));
+        _sensors.addPartitionSensors(partition);
+      }
+      _produceExecutor.scheduleWithFixedDelay(new ProduceRunnable(partition, key), _produceDelayMs, _produceDelayMs, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private Map<Integer, String> generateKeyMappings() {
+    int partitionNum = _partitionNum.get();
+    HashMap<Integer, String> keyMapping = new HashMap<>();
+
+    int keyIndex = 0;
+    while (keyMapping.size() < partitionNum) {
+      String key = keyIndex + "";
+      int partition = getDefaultPartitionForKey(key);
+      if (!keyMapping.containsKey(partition)) {
+        keyMapping.put(partition, key);
+      }
+      keyIndex++;
+    }
+
+    return keyMapping;
+  }
+
+  //TODO: Remove method and convert uses to Util.toPositive when kafka version dependency reaches 0.10.1.10
+  public static int toPositive(int number) {
+    return number & 0x7fffffff;
+  }
+
+  private int getDefaultPartitionForKey(String key) {
+    byte[] keyBytes = key.getBytes();
+    return toPositive(murmur2(keyBytes)) % _partitionNum.get();
+  }
+
+
+  /**
+   * This creates the records sent to the consumer.
+   */
+  private class ProduceRunnable implements Runnable {
+    private final int _partition;
+    private final String _key;
+
+    ProduceRunnable(int partition, String key) {
+      _partition = partition;
+      _key = key;
+    }
+
+    public void run() {
+      try {
+        long nextIndex = _nextIndexPerKey.get(_key).get();
+        String message = Utils.jsonFromFields(_topic, nextIndex, System.currentTimeMillis(), _producerId, _recordSize);
+        BaseProducerRecord record = new BaseProducerRecord(_topic, null, _key, message);
```

Comment:
Reuse would probably be higher if there was a super class with protected method byte[] getKey().  This can return null in the super class and in the keyed version it can return the key bytes.